pyccel/ast/builtins.py

```diff
@@ -513,11 +513,6 @@ def __init__(self, *args):
 
         # Get possible datatypes
         dtypes = [a.class_type.datatype for a in args]
-        # Extract all dtypes inside any inhomogeneous tuples
-        while any(isinstance(d, InhomogeneousTupleType) for d in dtypes):
-            dtypes = [di for d in dtypes for di in ((d_elem.datatype for d_elem in d)
-                                                    if isinstance(d, InhomogeneousTupleType)
-                                                    else [d])]
```

Comment:
Why this change?

Reply:
This is now handled inside `InhomogeneousTupleType` as it is a reasonably common request. I am going to make a minor modification though to calculate this in the constructor instead of the `datatype` method